File: tests/cpu/julia/test_audio.jl
# Audio/I/O CPU binding tests
# Run with:
#   LD_LIBRARY_PATH=build/backends/cpu julia tests/cpu/julia/test_audio.jl

push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "bindings", "julia"))
push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "build", "bindings", "julia"))

using Insight

passed = 0
failed = 0

function check(name, cond)
    global passed, failed
    if cond
        passed += 1
    else
        failed += 1
        println("FAIL: $name")
    end
end

println("=== Audio I/O ===")

# Note: Audio WAV read/write is C++ only (audio::read/audio::write).
# These tests verify signal I/O functions available in bindings.

tmpdir = mktempdir(prefix="insight_jl_audio_")

# write read roundtrip float64
try
    a = Insight.from_data([1.0, 2.5, 3.7, -1.2, 0.0])
    path = joinpath(tmpdir, "test_f64.bin")
    Insight.write_bin(path, a)
    result = Insight.read_bin(path, Insight.float64)
    check("write_read_f64", Insight.numel(result) == 5)
catch e
    println("SKIP: write_read_f64 ($e)")
end

# write read roundtrip float32
try
    a = Insight.zeros(Int64[4], Insight.float32)
    path = joinpath(tmpdir, "test_f32.bin")
    Insight.write_bin(path, a)
    result = Insight.read_bin(path, Insight.float32)
    check("write_read_f32", Insight.numel(result) == 4)
catch e
    println("SKIP: write_read_f32 ($e)")
end

# write read roundtrip int32
try
    a = Insight.from_data([10.0, 20.0, 30.0])
    path = joinpath(tmpdir, "test_i32.bin")
    Insight.write_bin(path, a)
    result = Insight.read_bin(path, Insight.int32)
    check("write_read_i32", Insight.numel(result) == 3)
catch e
    println("SKIP: write_read_i32 ($e)")
end

# write read zeros
try
    a = Insight.zeros(Int64[100], Insight.float64)
    path = joinpath(tmpdir, "test_zeros.bin")
    Insight.write_bin(path, a)
    result = Insight.read_bin(path, Insight.float64)
    check("write_read_zeros", Insight.numel(result) == 100)
catch e
    println("SKIP: write_read_zeros ($e)")
end

# write read ones
try
    a = Insight.ones(Int64[50], Insight.float64)
    path = joinpath(tmpdir, "test_ones.bin")
    Insight.write_bin(path, a)
    result = Insight.read_bin(path, Insight.float64)
    check("write_read_ones", Insight.numel(result) == 50)
catch e
    println("SKIP: write_read_ones ($e)")
end

# write read negative
try
    a = Insight.from_data([-100.5, -200.3, -300.1])
    path = joinpath(tmpdir, "test_neg.bin")
    Insight.write_bin(path, a)
    result = Insight.read_bin(path, Insight.float64)
    check("write_read_neg", Insight.numel(result) == 3)
catch e
    println("SKIP: write_read_neg ($e)")
end

# write read single
try
    a = Insight.from_data([42.0])
    path = joinpath(tmpdir, "test_single.bin")
    Insight.write_bin(path, a)
    result = Insight.read_bin(path, Insight.float64)
    d = Insight.to_array(result)
    check("write_read_single", Insight.numel(result) == 1 && isapprox(d[1], 42.0))
catch e
    println("SKIP: write_read_single ($e)")
end

# write read large
try
    t = Float64[i * 0.1 for i in 1:1000]
    a = Insight.from_data(t)
    path = joinpath(tmpdir, "test_large.bin")
    Insight.write_bin(path, a)
    result = Insight.read_bin(path, Insight.float64)
    check("write_read_large", Insight.numel(result) == 1000)
catch e
    println("SKIP: write_read_large ($e)")
end

# write creates file
try
    a = Insight.from_data([1.0, 2.0, 3.0])
    path = joinpath(tmpdir, "test_create.bin")
    Insight.write_bin(path, a)
    check("write_creates_file", isfile(path) && filesize(path) > 0)
catch e
    println("SKIP: write_creates_file ($e)")
end

# write overwrite
try
    a1 = Insight.from_data([1.0, 2.0, 3.0])
    a2 = Insight.from_data([4.0, 5.0, 6.0, 7.0])
    path = joinpath(tmpdir, "test_overwrite.bin")
    Insight.write_bin(path, a1)
    Insight.write_bin(path, a2)
    result = Insight.read_bin(path, Insight.float64)
    check("write_overwrite", Insight.numel(result) == 4)
catch e
    println("SKIP: write_overwrite ($e)")
end

# file size correct
try
    a = Insight.from_data([1.0, 2.0, 3.0, 4.0, 5.0])
    path = joinpath(tmpdir, "test_size.bin")
    Insight.write_bin(path, a)
    check("file_size", filesize(path) == 40)  # 5 * 8 bytes
catch e
    println("SKIP: file_size ($e)")
end

# ============================================================================
# Cleanup
# ============================================================================
try
    rm(tmpdir, recursive=true, force=true)
catch
end

# ============================================================================
# Results
# ============================================================================
println("\n" * "="^40)
println("Results: $passed passed, $failed failed")
if failed > 0
    exit(1)
end

File: tests/cpu/lua/test_audio.lua
-- Audio/I/O CPU binding tests
-- Run with:
--   LUA_PATH="bindings/lua/?/init.lua;;" LUA_CPATH="build/bindings/lua/?.so;;" \
--   LD_LIBRARY_PATH=build/backends/cpu ~/.luarocks/bin/busted tests/cpu/lua/test_audio.lua

describe("Audio CPU Tests", function()
  local ins
  setup(function()
    ins = require("_insight")
  end)

  -- Note: Audio WAV read/write is C++ only (audio::read/audio::write).
  -- These tests verify signal I/O functions available in bindings.

  it("write creates file", function()
    local a = ins.from_table({ 1.0, 2.0, 3.0 })
    local path = "/tmp/insight_lua_audio_test.bin"
    ins.signal.write_bin(path, a)
    local f = io.open(path, "r")
    assert.is_not_nil(f)
    f:close()
    os.remove(path)
  end)

  it("write read roundtrip float64", function()
    local a = ins.from_table({ 1.0, 2.5, 3.7, -1.2, 0.0 })
    local path = "/tmp/insight_lua_audio_f64.bin"
    ins.signal.write_bin(path, a)
    local result = ins.signal.read_bin(path, ins.float64)
    assert.is_not_nil(result)
    assert.are.equal(5, result.numel)
    os.remove(path)
  end)

  it("write read roundtrip float32", function()
    local a = ins.zeros({ 4 }, ins.float32)
    local path = "/tmp/insight_lua_audio_f32.bin"
    ins.signal.write_bin(path, a)
    local result = ins.signal.read_bin(path, ins.float32)
    assert.is_not_nil(result)
    assert.are.equal(4, result.numel)
    os.remove(path)
  end)

  it("write read roundtrip int32", function()
    local a = ins.from_table({ 10.0, 20.0, 30.0 })
    a = ins.cast(a, ins.int32)
    local path = "/tmp/insight_lua_audio_i32.bin"
    ins.signal.write_bin(path, a)
    local result = ins.signal.read_bin(path, ins.int32)
    assert.is_not_nil(result)
    assert.are.equal(3, result.numel)
    os.remove(path)
  end)

  it("write read zeros", function()
    local a = ins.zeros({ 100 }, ins.float64)
    local path = "/tmp/insight_lua_audio_zeros.bin"
    ins.signal.write_bin(path, a)
    local result = ins.signal.read_bin(path, ins.float64)
    assert.is_not_nil(result)
    assert.are.equal(100, result.numel)
    os.remove(path)
  end)

  it("write read ones", function()
    local a = ins.ones({ 50 }, ins.float64)
    local path = "/tmp/insight_lua_audio_ones.bin"
    ins.signal.write_bin(path, a)
    local result = ins.signal.read_bin(path, ins.float64)
    assert.is_not_nil(result)
    assert.are.equal(50, result.numel)
    os.remove(path)
  end)

  it("write read negative values", function()
    local a = ins.from_table({ -100.5, -200.3, -300.1 })
    local path = "/tmp/insight_lua_audio_neg.bin"
    ins.signal.write_bin(path, a)
    local result = ins.signal.read_bin(path, ins.float64)
    assert.is_not_nil(result)
    assert.are.equal(3, result.numel)
    os.remove(path)
  end)

  it("write read single element", function()
    local a = ins.from_table({ 42.0 })
    local path = "/tmp/insight_lua_audio_single.bin"
    ins.signal.write_bin(path, a)
    local result = ins.signal.read_bin(path, ins.float64)
    assert.is_not_nil(result)
    assert.are.equal(1, result.numel)
    os.remove(path)
  end)

  it("write read large signal", function()
    local t = {}
    for i = 1, 1000 do
      t[i] = i * 0.1
    end
    local a = ins.from_table(t)
    local path = "/tmp/insight_lua_audio_large.bin"
    ins.signal.write_bin(path, a)
    local result = ins.signal.read_bin(path, ins.float64)
    assert.is_not_nil(result)
    assert.are.equal(1000, result.numel)
    os.remove(path)
  end)

  it("write file size correct", function()
    local a = ins.from_table({ 1.0, 2.0, 3.0, 4.0, 5.0 })
    local path = "/tmp/insight_lua_audio_size.bin"
    ins.signal.write_bin(path, a)
    local f = io.open(path, "r")
    local size = f:seek("end")
    f:close()
    assert.are.equal(40, size) -- 5 doubles * 8 bytes
    os.remove(path)
  end)

  it("write read overwrite", function()
    local a1 = ins.from_table({ 1.0, 2.0, 3.0 })
    local a2 = ins.from_table({ 4.0, 5.0, 6.0, 7.0 })
    local path = "/tmp/insight_lua_audio_overwrite.bin"
    os.remove(path)
    ins.signal.write_bin(path, a1, false)
    ins.signal.write_bin(path, a2, false)
    local result = ins.signal.read_bin(path, ins.float64)
    assert.is_not_nil(result)
    assert.are.equal(4, result.numel)
    os.remove(path)
  end)
end)
